### ISSUE TITLE
ICE bootstrapping rustc with NLL enabled

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#48071]
When I try to bootstrap rustc with NLL enabled, I get an odd ICE compiling `libsyntax_ext`:

```
error: internal compiler error: broken MIR in NodeId(16117) (_368 = move _369): bad assignment (for<'cx, 'r, 's> fn(&'cx mut syntax::ext::base::ExtCtxt<'r>, syntax_pos::Span, &'s [syntax::tokenstream::To\
kenTree]) -> std::boxed::Box<syntax::ext::base::MacResult + 'cx> = for<'r, 's, 't0> fn(&'r mut syntax::ext::base::ExtCtxt<'s>, syntax_pos::Span, &'t0 [syntax::tokenstream::TokenTree]) -> std::boxed::Box<\
syntax::ext::base::MacResult>): RegionsOverlyPolymorphic(BrNamed(crate11:DefIndex(1:1012), 'cx), '_#729r)
   --> src/libsyntax_ext/lib.rs:65:44
    |
65  |                           expander: Box::new($f as MacroExpanderFn),
    |                                              ^^^^^^^^^^^^^^^^^^^^^
...
92  | /     register! {
93  | |         line: expand_line,
94  | |         __rust_unstable_column: expand_column_gated,
95  | |         column: expand_column,
...   |
112 | |         compile_error: compile_error::expand_compile_error,
113 | |     }
    | |_____- in this macro invocation
```

**Comment by Aaron1011** [comment]
I'd like to work on this.

**Comment by nikomatsakis** [comment]
@Aaron1011 any progress? want me to take a look at all?

**Comment by Aaron1011** [comment]
@nikomatsakis: I haven't yet been able to reproduce this, due to https://github.com/rust-lang/rust/issues/48129 stopping compilation. I'll try to work around that issue locally, and see if I can reproduce + investigate the original issue.

**Comment by Aaron1011** [comment]
@nikomatsakis: I ran across another issue while trying to reach the ICE: https://github.com/rust-lang/rust/issues/48224

**Comment by Aaron1011** [comment]
@nikomatsakis: Until the intermediate issues are resolved, I don't think I'll be able to reproduce this.

**Comment by nikomatsakis** [comment]
@Aaron1011 this branch https://github.com/rust-lang/rust/pull/48372 clears the way

**Comment by nikomatsakis** [comment]
Hmm, but on that branch I am now encountering the `AddAssign` problems I guess. So we're still blocked on #48129

**Comment by nikomatsakis** [comment]
Actually, the issue-48071 branch on my repo reproduces the problem. You have to build like so:

```
RUSTFLAGS_STAGE_NOT_0='-Znll -Zborrowck=mir -Ztwo-phase-borrows' ./x.py build --stage 2 
```

**Comment by nikomatsakis** [comment]
Error that I get is:

```
error: internal compiler error: broken MIR in NodeId(16117) (_368 = move _369): bad assignment (

for<'cx, 'r, 's> fn(&'cx mut syntax::ext::base::ExtCtxt<'r>, syntax_pos::Span, &'s [syntax::tokenstream::TokenTree]) -> std::boxed::Box<syntax::ext::base::MacResult + 'cx>
for<'r, 's, 't0> fn(&'r mut syntax::ext::base::ExtCtxt<'s>, syntax_pos::Span, &'t0 [syntax::tokenstream::TokenTree]) -> std::boxed::Box<syntax::ext::base::MacResult>): 

RegionsOverlyPolymorphic(BrNamed(crate11:DefIndex(1:1012), 'cx), '_#729r)
```

**Comment by nikomatsakis** [comment]
Minimized:

```rust
#![allow(warnings)]
#![feature(dyn_trait)]
#![feature(nll)]

trait Foo {
    fn foo(&self) { }
}

impl Foo for () {
}

type MakeFooFn = for<'a> fn(&'a u8) -> Box<dyn Foo + 'a>;

fn make_foo(x: &u8) -> Box<dyn Foo + 'static> {
    Box::new(())
}

fn main() {
    let x: MakeFooFn = make_foo as MakeFooFn;
}
```

**Comment by nikomatsakis** [comment]
I think the problem is that there is a missing coercion of sorts, or else that this code in the NLL type checker is too strict:

https://github.com/rust-lang/rust/blob/27a046e9338fb0455c33b13e8fe28da78212dedc/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L1308-L1330

In particular, it constructs the "reified" form of the fn pointer in `ty_fn_ptr_from` and then requires that the target type be *equal* to that. But in this case, the target type is a *subtype* of that.

**Comment by nikomatsakis** [comment]
(Hmm, I have a vague memory of solving a similar problem somewhere else, where we were rolling multiple coercions into one. It's sort of debatable what's right here, I suppose -- probably we should change the NLL type checker to use `sub_types`.)

**Comment by nikomatsakis** [comment]
OK well that proposed change didn't actually fix anything =)

**Comment by nikomatsakis** [comment]
Oh, interesting. The problem, I believe, is https://github.com/rust-lang/rust/issues/33684 -- which @sgrif and I have been working on a PR to fix, independently.

**Comment by nikomatsakis** [comment]
Tagging as NLL-deferred -- this is blocked on @sgrif's work at the moment and there exists a workaround for within rustc.

**Comment by nikomatsakis** [comment]
Hmm. I'm going to mark this as high priority because it blocks bootstrap. I think it's a bit tricky to fix. While bike riding recently though I had some ideas here -- I realized that we could probably write a custom bit of subtyping code used only (initially) by the MIR type checker. The hope would be that this is faster than canonicalizing etc. This code could use a universe-based check. This would sidestep the limitations of our existing subtyping algorithm which are biting us here. It would also help eliminate a blocker for chalk integration later. I may take a stab at this.

**Comment by pnkfelix** [comment]
visited for triage. this should be fixed via PR #52488 (which is WIP but hopefully will be ready soon).
